src/main/java/dev/openfeature/sdk/ImmutableContext.java

```diff
@@ -86,10 +87,34 @@ public EvaluationContext merge(EvaluationContext overridingContext) {
         if (overridingContext.getTargetingKey() != null && !overridingContext.getTargetingKey().trim().equals("")) {
             newTargetingKey = overridingContext.getTargetingKey();
         }
-        Map<String, Value> merged = new HashMap<>();
 
-        merged.putAll(this.asMap());
-        merged.putAll(overridingContext.asMap());
+        Map<String, Value> merged = this.merge(this.asMap(), overridingContext.asMap());
+
         return new ImmutableContext(newTargetingKey, merged);
     }
+
+    /**
+     * Recursively merges the base map from this EvaluationContext with the passed EvaluationContext.
+     * 
+     * @param base base map to merge
+     * @param overriding overriding map to merge
+     * @return resulting merged map
+     */
+    private Map<String, Value> merge(Map<String, Value> base, Map<String, Value> overriding) {
```

Comment:
Both this, and the similar method in `MutableContext` seem very similar. I think you can extract them both to a common protected default implementation, and somehow extract or economize the instantiation [here](https://github.com/open-feature/java-sdk/pull/280/files#diff-3b1bb0fba16899446d6f2ef99d562c9072f6344156ed7f243910bb8fa27d8f57R113), which seems to be the only difference. You could perhaps use an abstract method for returning a new instance, or pass a lambda or something like that.

Reply:
Yes, this is correct & perfect suggestion.

What differ is the context type (mutable vs immutable). This can be easily handled with a functional interface [1]. And I would see this best fitting as a default function exposed from `Structure.java`

In pseudocode, 

```
// Structure.java

default <T extends Structure> void merge(Function<String, T> newFunc, ......args ){

   newStructure.apply(params);

}
```


[1] - https://docs.oracle.com/javase/8/docs/api/java/util/function/Function.html

Reply:
I would say the default method with the recursive merge should be in `EvaluationContext.java` because `Structure` doesn't have `merge`.

Reply:
I think the recursive merge is actually merging the two structure. so I would also prefer to keeping it in the `Structure.java`.

Reply:
Moved to `Structure.java` with a `Function<Map<String,Value>, T>` as first parameter.